### ISSUE TITLE
OutOfMemoryError for some images

### ISSUE THREAD

**Issue by dusan-rychnovsky** [drewnoakes/metadata-extractor#167]
For some images `ImageMetadataReader#readMetadata(InputStream inputStream)` crashes with an `OutOfMemoryError` with the following stack trace.

```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at com.drew.imaging.tiff.TiffReader.processTag(TiffReader.java:369)
	at com.drew.imaging.tiff.TiffReader.processIfd(TiffReader.java:228)
	at com.drew.imaging.tiff.TiffReader.processIfd(TiffReader.java:248)
	at com.drew.imaging.tiff.TiffReader.processTiff(TiffReader.java:78)
	at com.drew.metadata.exif.ExifReader.extract(ExifReader.java:89)
	at com.drew.metadata.exif.ExifReader.readJpegSegments(ExifReader.java:74)
	at com.drew.imaging.jpeg.JpegMetadataReader.processJpegSegmentData(JpegMetadataReader.java:130)
	at com.drew.imaging.jpeg.JpegMetadataReader.process(JpegMetadataReader.java:122)
	at com.drew.imaging.jpeg.JpegMetadataReader.readMetadata(JpegMetadataReader.java:73)
	at com.drew.imaging.jpeg.JpegMetadataReader.readMetadata(JpegMetadataReader.java:80)
	at com.drew.imaging.ImageMetadataReader.readMetadata(ImageMetadataReader.java:112)
	at com.drew.imaging.ImageMetadataReader.readMetadata(ImageMetadataReader.java:90)
	...
```

An example would be this image http://www.pgs-golfers.com/images/IMG_1739.jpg where it crashes when processing tag with `tagId == 514`.

This happens despite the fact that there is the following guard against this situation in `TiffReader` on lines 206 - 211:

```
// Check that this tag isn't going to allocate outside the bounds of the data array.
// This addresses an uncommon OutOfMemoryError.
if (byteCount < 0 || tagValueOffset + byteCount > reader.getLength()) {
  handler.error("Illegal number of bytes for TIFF tag data: " + byteCount);
  continue;
}
```

The problem is that `byteCount` is declared as `int` and since for this tag `componentCount == 1073741825`, the following line causes an arithmetic overflow:

```
byteCount = componentCount * format.getComponentSizeBytes();
```

and thus `byteCount == 4` instead of `byteCount == 4294967300`.

**Comment by drewnoakes** [comment]
Are you able to donate that image to the sample library for regression testing?

Want to make a PR for this?

**Comment by dusan-rychnovsky** [comment]
I don't have any rights to that image - it's someone else's image exposed on that URL. Is it ok to upload it to this GitHub repository?

Yes I can make a pull request.

**Comment by drewnoakes** [comment]
The owner of the image would need to be willing to donate it for testing
purposes. There's another repository here with more information:

https://github.com/drewnoakes/metadata-extractor-images

**Comment by dusan-rychnovsky** [comment]
Should I send a pull request with a fix without a test case?

**Comment by drewnoakes** [comment]
It should be reasonably straightforward to make a test case from a byte
array rather than an image. You'd only need a TIFF header, then an IFD with
a single tag exhibiting the out of bounds case when overflow is fixed, but
that is in bounds if the byte count overflows. You can pad the end of the
array with zeroes if needed.

**Comment by drewnoakes** [comment]
I've patched this on my local PC. Will get it pushed and released soon.

**Comment by dusan-rychnovsky** [comment]
ok, great :) thank you

**Comment by drewnoakes** [comment]
Just released this in 2.9.1.
